File: CS/ViewModel/MainViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Threading.Tasks;
using System.Windows;

namespace Example.ViewModel {
    public class MainViewModel : ViewModelBase {
        RegistrationViewModel registrationViewModel;

        IMessageBoxService MessageBoxService { get { return GetService<IMessageBoxService>(); } }
        IDialogService DialogService { get { return GetService<IDialogService>(); } }

        public MainViewModel() {
            registrationViewModel = new RegistrationViewModel();
        }

        [Command]
        public void ShowRegistrationForm() {
            UICommand registerAsyncCommand = new UICommand(
               id: null,
               caption: "Register Async",
               command: new AsyncCommand<CancelEventArgs>(
                   async cancelArgs => {
                       try {
                           await MyAsyncExecuteMethod();
                       }
                       catch (Exception e) {
                           MessageBoxService.ShowMessage(e.Message, "Error", MessageButton.OK);
                           cancelArgs.Cancel = true;
                       }
                   },
                   cancelArgs => !string.IsNullOrEmpty(registrationViewModel.UserName)
               ),
               asyncDisplayMode: AsyncDisplayMode.Wait,
               isDefault: false,
               isCancel: false
           );

            UICommand registerCommand = new UICommand(
                id: null,
                caption: "Register",
                command: new DelegateCommand<CancelEventArgs>(
                    cancelArgs => {
                        try {
                            MyExecuteMethod();
                        }
                        catch (Exception e) {
                            MessageBoxService.ShowMessage(e.Message, "Error", MessageButton.OK);
                            cancelArgs.Cancel = true;
                        }
                    },
                    cancelArgs => !string.IsNullOrEmpty(registrationViewModel.UserName) && !((IAsyncCommand)registerAsyncCommand.Command).IsExecuting
                ),
                isDefault: true,
                isCancel: false
            );

            UICommand cancelCommand = new UICommand(
                id: MessageBoxResult.Cancel,
                caption: "Cancel",
                command: null,
                isDefault: false,
                isCancel: true
            );

            UICommand result = DialogService.ShowDialog(
                dialogCommands: new List<UICommand>() { registerAsyncCommand, registerCommand, cancelCommand },
                title: "Registration Dialog",
                viewModel: registrationViewModel
            );

            if (result == registerCommand) {
                // ...
            }
        }

        void MyExecuteMethod() {
            // ...
        }

        async Task MyAsyncExecuteMethod() {
            await Task.Delay(2000);
            // ...
        }
    }
}

File: VB/ViewModel/MainViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Windows

Namespace Example.ViewModel

    Public Class MainViewModel

        Protected ReadOnly Property DialogService As IDialogService
            Get
                Return GetService(Of IDialogService)()
            End Get
        End Property

        Private registrationViewModel As RegistrationViewModel = Nothing

        Public Sub ShowRegistrationForm()
            If registrationViewModel Is Nothing Then registrationViewModel = New RegistrationViewModel()
            Dim registerCommand As UICommand = New UICommand() With {.Caption = "Register", .IsCancel = False, .IsDefault = True, .Command = New DelegateCommand(Of CancelEventArgs)(Sub(x) myExecuteMethod(), Function(x) Not String.IsNullOrEmpty(registrationViewModel.UserName))}
            Dim cancelCommand As UICommand = New UICommand() With {.Id = MessageBoxResult.Cancel, .Caption = "Cancel", .IsCancel = True, .IsDefault = False}
            Dim result As UICommand = DialogService.ShowDialog(dialogCommands:=New List(Of UICommand)() From {registerCommand, cancelCommand}, title:="Registration Dialog", viewModel:=registrationViewModel)
            If result Is registerCommand Then
            '...
            End If
        End Sub

        Private Sub myExecuteMethod()
            Call Debug.Print("Registration complete")
        End Sub
    End Class
End Namespace
